src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/BufferBlockCopyLengthAnalyzer.cs

```diff
@@ -0,0 +1,190 @@
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                if (!compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemBuffer, out INamedTypeSymbol? bufferType))
+                {
+                    return;
+                }
+
+                IMethodSymbol blockCopyMethod = bufferType
+                   .GetMembers("BlockCopy")
+                   .OfType<IMethodSymbol>()
+                   .FirstOrDefault();
+
+                if (blockCopyMethod is null)
+                {
+                    return;
+                }
+
+                if (!compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemArray, out INamedTypeSymbol? arrayType))
+                {
+                    return;
+                }
+
+                IPropertySymbol arrayLengthProp = arrayType
+                    .GetMembers("Length")
+                    .OfType<IPropertySymbol>()
+                    .FirstOrDefault();
+
+                if (arrayLengthProp is null)
+                {
+                    return;
+                }
+
+                if (!compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemByte, out INamedTypeSymbol? byteType))
+                {
+                    return;
+                }
+
+                if (!compilationContext.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemSByte, out INamedTypeSymbol? sByteType))
+                {
+                    return;
+                }
+
+                compilationContext.RegisterOperationAction(context =>
+                {
+                    var invocationOperation = (IInvocationOperation)context.Operation;
+                    if (invocationOperation.Arguments.IsEmpty)
+                    {
+                        return;
+                    }
+
+                    if (!invocationOperation.TargetMethod.Equals(blockCopyMethod))
+                    {
+                        return;
+                    }
+
+                    ImmutableArray<IArgumentOperation> arguments = invocationOperation.Arguments;
+                    if (arguments.Length != 5)
+                    {
+                        return;
+                    }
+
+                    IArgumentOperation srcArg = arguments[0];
+                    IArgumentOperation dstArg = arguments[2];
+                    IArgumentOperation lastArg = arguments[4];
+
+                    bool CheckLengthProperty(IPropertyReferenceOperation lastArg)
+                    {
+                        if (lastArg.Property.Equals(arrayLengthProp))
+                        {
+                            if (srcArg.Value is IConversionOperation srcArgValue)
+                            {
+                                if (lastArg.Instance.GetReferencedMemberOrLocalOrParameter() == srcArgValue.Operand.GetReferencedMemberOrLocalOrParameter())
```

Comment:
The lines inside this if block are repeated for dstArg. I would move these to another local function and call it from both places